File: python-service/app/parsers/llm_parser.py
"""LLM-powered address parser that extracts structured address fields from raw text."""

import asyncio
import json
from typing import Any, Dict

import httpx

from app.models.address import StructuredAddress


class LLMTimeoutError(Exception):
    """Raised when the LLM service does not respond within the configured timeout."""

    def __init__(self, timeout: float):
        self.timeout = timeout
        super().__init__(
            f"LLM service did not respond within {timeout} seconds"
        )


class LLMParseError(Exception):
    """Raised when the LLM response cannot be parsed into a valid StructuredAddress."""

    def __init__(self, raw_text: str, reason: str = ""):
        self.raw_text = raw_text
        self.reason = reason
        message = f"Failed to parse address from text: {raw_text!r}"
        if reason:
            message += f" — {reason}"
        super().__init__(message)


class LLMParser:
    """Parses unstructured address text into structured fields using an LLM service."""

    def __init__(self, endpoint_url: str = "", api_key: str = ""):
        self.endpoint_url = endpoint_url
        self.api_key = api_key

    async def parse(self, raw_text: str, timeout: float = 10.0) -> StructuredAddress:
        """Parse unstructured address text into structured fields using LLM.

        Args:
            raw_text: The raw, unstructured address text to parse.
            timeout: Maximum seconds to wait for the LLM service response.
                     Defaults to 10.0 seconds.

        Returns:
            A StructuredAddress extracted from the raw text.

        Raises:
            LLMTimeoutError: If the LLM service doesn't respond within timeout.
            LLMParseError: If the LLM response cannot be converted to a valid
                           StructuredAddress.
        """
        try:
            response = await asyncio.wait_for(
                self._call_llm_service(raw_text),
                timeout=timeout,
            )
        except asyncio.TimeoutError:
            raise LLMTimeoutError(timeout=timeout)
        except httpx.TimeoutException:
            raise LLMTimeoutError(timeout=timeout)

        return self._extract_address(response, raw_text)

    async def _call_llm_service(self, raw_text: str) -> Dict[str, Any]:
        """Make the actual HTTP call to the LLM service.

        This method contains the real HTTP integration point. In production,
        it sends the raw text to the configured LLM endpoint and returns the
        parsed JSON response.

        Args:
            raw_text: The raw address text to send to the LLM.

        Returns:
            The JSON response from the LLM service as a dictionary.

        Raises:
            NotImplementedError: When no endpoint URL is configured (development mode).
            httpx.TimeoutException: When the HTTP request times out.
            LLMParseError: When the service returns an error or non-JSON response.
        """
        if not self.endpoint_url:
            raise NotImplementedError(
                "LLM service endpoint not configured. "
                "Set endpoint_url to enable LLM-powered parsing."
            )

        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        payload = {
            "prompt": self._build_prompt(raw_text),
            "raw_text": raw_text,
        }

        async with httpx.AsyncClient() as client:
            http_response = await client.post(
                self.endpoint_url,
                json=payload,
                headers=headers,
                timeout=None,  # Timeout is managed by asyncio.wait_for
            )

        if http_response.status_code != 200:
            raise LLMParseError(
                raw_text=raw_text,
                reason=f"LLM service returned status {http_response.status_code}",
            )

        try:
            return http_response.json()
        except (json.JSONDecodeError, ValueError):
            raise LLMParseError(
                raw_text=raw_text,
                reason="LLM service returned non-JSON response",
            )

    def _build_prompt(self, raw_text: str) -> str:
        """Build the prompt to send to the LLM for address extraction.

        Args:
            raw_text: The raw address text.

        Returns:
            A formatted prompt string.
        """
        return (
            "Extract the US address components from the following text. "
            "Return a JSON object with these exact keys: "
            "street_line_1, street_line_2, city, state, zipcode. "
            "If street_line_2 is not present, set it to null.\n\n"
            f"Text: {raw_text}"
        )

    def _extract_address(
        self, response: Dict[str, Any], raw_text: str
    ) -> StructuredAddress:
        """Extract a StructuredAddress from the LLM response.

        Args:
            response: The parsed JSON response from the LLM service.
            raw_text: The original raw text (for error reporting).

        Returns:
            A validated StructuredAddress.

        Raises:
            LLMParseError: If the response doesn't contain valid address fields.
        """
        # The LLM response may contain the address fields directly or nested
        address_data = response.get("address", response)

        required_fields = ["street_line_1", "city", "state", "zipcode"]
        missing_fields = [
            field for field in required_fields if not address_data.get(field)
        ]

        if missing_fields:
            raise LLMParseError(
                raw_text=raw_text,
                reason=f"Missing required fields: {', '.join(missing_fields)}",
            )

        try:
            return StructuredAddress(
                street_line_1=address_data["street_line_1"],
                street_line_2=address_data.get("street_line_2"),
                city=address_data["city"],
                state=address_data["state"],
                zipcode=address_data["zipcode"],
            )
        except (KeyError, TypeError, ValueError) as exc:
            raise LLMParseError(
                raw_text=raw_text,
                reason=f"Invalid address data in LLM response: {exc}",
            )

File: express-service/src/parsers/llmParser.ts
import axios, { AxiosError } from 'axios';
import { StructuredAddress } from '../models/address';

/**
 * Custom error thrown when the LLM service times out.
 */
export class LLMTimeoutError extends Error {
  constructor(timeoutMs: number) {
    super(`LLM service timed out after ${timeoutMs}ms`);
    this.name = 'LLMTimeoutError';
  }
}

/**
 * Custom error thrown when the LLM fails to parse the address text.
 */
export class LLMParseError extends Error {
  public readonly rawText: string;

  constructor(rawText: string, reason?: string) {
    super(reason ?? `Failed to parse address from text: "${rawText}"`);
    this.name = 'LLMParseError';
    this.rawText = rawText;
  }
}

const DEFAULT_TIMEOUT_MS = 10000;
const DEFAULT_LLM_ENDPOINT = process.env.LLM_ENDPOINT || 'http://localhost:8081/parse-address';

/**
 * LLM-powered address parser that extracts structured address fields
 * from unstructured or messy address text.
 *
 * Accepts a configurable endpoint URL for testability.
 */
export class LLMParser {
  private readonly endpoint: string;

  /**
   * @param endpoint - The LLM service endpoint URL. Defaults to LLM_ENDPOINT env var or http://localhost:8081/parse-address
   */
  constructor(endpoint?: string) {
    this.endpoint = endpoint ?? DEFAULT_LLM_ENDPOINT;
  }

  /**
   * Parse unstructured address text into a StructuredAddress using an LLM service.
   *
   * @param rawText - The unstructured address text to parse
   * @param timeout - Timeout in milliseconds (default: 10000ms / 10 seconds)
   * @returns A StructuredAddress extracted from the raw text
   * @throws LLMTimeoutError if the LLM service does not respond within the timeout
   * @throws LLMParseError if the LLM response cannot be parsed into a valid address
   */
  async parse(rawText: string, timeout: number = DEFAULT_TIMEOUT_MS): Promise<StructuredAddress> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const result = await this.callLLMService(rawText, controller.signal, timeout);
      clearTimeout(timeoutId);
      return this.extractAddress(result, rawText);
    } catch (error: unknown) {
      clearTimeout(timeoutId);

      if (error instanceof LLMTimeoutError || error instanceof LLMParseError) {
        throw error;
      }

      if (error instanceof Error && error.name === 'AbortError') {
        throw new LLMTimeoutError(timeout);
      }

      throw new LLMParseError(rawText, `LLM service error: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Call the external LLM service to parse address text.
   * Uses axios with timeout and abort signal support.
   */
  protected async callLLMService(rawText: string, signal: AbortSignal, timeout: number): Promise<string> {
    // Check if already aborted before proceeding
    if (signal.aborted) {
      const abortError = new Error('Aborted');
      abortError.name = 'AbortError';
      throw abortError;
    }

    try {
      const response = await axios.post(
        this.endpoint,
        { raw_address: rawText },
        {
          timeout,
          signal,
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );

      // Return the response data as a JSON string for extractAddress to parse
      if (typeof response.data === 'string') {
        return response.data;
      }
      return JSON.stringify(response.data);
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;

        // Timeout or cancelled request → service unavailable
        if (axiosError.code === 'ECONNABORTED' || axiosError.code === 'ERR_CANCELED') {
          throw new LLMTimeoutError(timeout);
        }

        // Network errors (connection refused, DNS failure, etc.) → service unavailable
        if (axiosError.code === 'ECONNREFUSED' || axiosError.code === 'ENOTFOUND' || !axiosError.response) {
          throw new LLMTimeoutError(timeout);
        }

        // Server returned an error response → parse failed
        throw new LLMParseError(rawText, `LLM service returned error: ${axiosError.response?.status ?? 'unknown'}`);
      }

      // AbortError from signal
      if (error instanceof Error && error.name === 'AbortError') {
        throw error;
      }

      throw error;
    }
  }

  /**
   * Extract a StructuredAddress from the LLM response text.
   * Expects the LLM to return JSON with address fields.
   */
  private extractAddress(llmResponse: string, rawText: string): StructuredAddress {
    try {
      const parsed = JSON.parse(llmResponse);

      if (!parsed.street_line_1 || !parsed.city || !parsed.state || !parsed.zipcode) {
        throw new LLMParseError(rawText, 'LLM response missing required address fields');
      }

      return {
        street_line_1: parsed.street_line_1,
        ...(parsed.street_line_2 ? { street_line_2: parsed.street_line_2 } : {}),
        city: parsed.city,
        state: parsed.state,
        zipcode: parsed.zipcode,
      };
    } catch (error: unknown) {
      if (error instanceof LLMParseError) {
        throw error;
      }
      throw new LLMParseError(rawText, 'Failed to parse LLM response as structured address');
    }
  }
}
